pandas/_libs/index_class_helper.pxi.in

```diff
@@ -36,10 +46,33 @@ cdef class {{name}}Engine(IndexEngine):
     cdef _make_hash_table(self, n):
         {{if name == 'Object'}}
         return _hash.PyObjectHashTable(n)
+        {{elif name in {'Int8', 'Int16', 'Int32'} }}
+        # {{name}}HashTable is not available, so we use Int64HashTable
+        return _hash.Int64HashTable(n)
```

Comment:
Since all the Int8Engine, Int16Engine, .. are actually still using the same Int64HashTable, is it actually needed to create those different engines? Can the `Int64Engine` (or IntEngine) not simply always do the `ensure_int64` in `_call_map_locations`? (which seems to be the only difference between Int8Engine and Int64Engine, from a quick skim, didn't look in detail)

Reply:
The requirement needed to achieve the speedup is avoiding type conversion. I'm pretty sure that ``ensure_int64`` will convert a int8 array, so lose speeed? 

I'm not at the coding computer ATM, but I can double check how ``ensure_int64`` works tonight.

Reply:
> I'm pretty sure that ensure_int64 will convert a int8 array to int64, so lose speeed?

But the Int8Engine, Int16Engine, .. are actually calling `ensure_int64` in the current PR ? (not `ensure_int8`, ...)

Reply:
the ``self.mapping.map_locations(algos.ensure_int64(values))`` is only converting *``values``*, i.e. typically a scalar, so it's very fast.

It would be a different an dslower matter if we had converted the array, but we don't do that...

Reply:
Ah, OK, I see in `_get_index_values` it is `ensure_{{dtype}}` that is used.